[staging] ninja: 1.11.1 -> 1.12.1

JohnRTitor opened the pull request:
## Description of changes
- Move `ninja` to `pkgs/by-name`
- Changelog: https://github.com/ninja-build/ninja/compare/v1.11.1...v1.12.1
- Remove patch large fs for since it is merged to v1.12.0 (https://github.com/ninja-build/ninja/pull/2203)
- Closes https://github.com/NixOS/nixpkgs/issues/303406

Current meson version can not be built using ninja 1.12, so use a patch from upstream (https://github.com/mesonbuild/meson/commit/1b9eb6f3134d61c5ea0026ac95d807aab812c414) to allow building meson.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).


---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

JohnRTitor committed:
ninja: 1.11.1 -> 1.12.0

Remove patch for files in large fs, since it is merged to v1.12.0

JohnRTitor commented:
@ofborg build ninja

JohnRTitor commented:
@ofborg build ninja

JohnRTitor commented:
@ofborg build ninja

JohnRTitor commented:
@ofborg build pkgsi686Linux.ninja

JohnRTitor commented:
@ofborg build ninja

JohnRTitor commented:
@ofborg build pkgsi686Linux.ninja

JohnRTitor commented:
@ofborg build ninja

JohnRTitor commented:
@ofborg build ninja

JohnRTitor commented:
@ofborg build ninja

NickCao commented on pkgs/development/tools/build-managers/ninja/default.nix:
```diff
@@ -62,6 +48,10 @@ stdenv.mkDerivation rec {
     CFLAGS="$CFLAGS_FOR_BUILD" \
     CXXFLAGS="$CXXFLAGS_FOR_BUILD" \
     LDFLAGS="$LDFLAGS_FOR_BUILD" \
+
+    # write rebuild args to file after bootstrap
+    substituteInPlace configure.py --replace "subprocess.check_call(rebuild_args)" "open('rebuild_args','w').write(rebuild_args[0])"
+
```
Why this change.

JohnRTitor replied:
The *patch* is not applied anymore, so it makes sense to move the snippet from postPatch to buildPhase. 

Additionally ninja can not build at all without this substituteInPlace command, even with or without the patch applied.

NickCao replied:
`postPatch` has nothing to do with patches, you can leave it there.

JohnRTitor replied:
I've updated it, please check.

NickCao replied:
> I've updated it, please check.

The version update is accidentally dropped?

JohnRTitor replied:
Thanks for noticing.

JohnRTitor commented:
@ofborg build ninja

JohnRTitor commented:
@ofborg build ninja

JohnRTitor commented:
@NickCao ofborg eval is failing, even after multiple rebase. I can't seem to figure out what's wrong from the logs. I can successfully build locally and using ofborg here though.

NickCao reviewed (commented):
Skimmed through upstream diff, LGTM. Running nixpkgs-review.

NickCao commented:
@ofborg eval

NickCao commented:
This seems to somehow break meson.

NickCao commented:
> This seems to somehow break meson.

Should be fixed by https://github.com/mesonbuild/meson/commit/1b9eb6f3134d61c5ea0026ac95d807aab812c414

JohnRTitor commented:
> Should be fixed by [mesonbuild/meson@1b9eb6f](https://github.com/mesonbuild/meson/commit/1b9eb6f3134d61c5ea0026ac95d807aab812c414)

Should I make a PR to fetch the referenced commit as a patch to `meson`? Should allow building it I believe.

NickCao commented:
> Should I make a PR to fetch the referenced commit as a patch to `meson`? Should allow building it I believe.

Just add that to this PR.

JohnRTitor commented:
Looks like coreutils is failing at checkPhase.
```
============================================================================
Testsuite summary for GNU coreutils 9.5
============================================================================
# TOTAL: 653
# PASS:  486
# SKIP:  166
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See ./tests/test-suite.log
Please report to bug-coreutils@gnu.org
============================================================================
make[4]: *** [Makefile:22562: tests/test-suite.log] Error 1
make[4]: Leaving directory '/build/coreutils-9.5'
make[3]: *** [Makefile:22670: check-TESTS] Error 2
make[3]: Leaving directory '/build/coreutils-9.5'
make[2]: *** [Makefile:22930: check-am] Error 2
make[2]: Leaving directory '/build/coreutils-9.5'
make[1]: *** [Makefile:22447: check-recursive] Error 1
make[1]: Leaving directory '/build/coreutils-9.5'
make: *** [Makefile:22932: check] Error 2
```

JohnRTitor commented:
Likely because of recent update to 9.5. https://github.com/NixOS/nixpkgs/pull/299945

Pinging @LeSuisse @dasJ

LeSuisse commented:
Hard to tell from just this sumary but if this a Musl & coreutils thing see #304856

JohnRTitor commented:
@ofborg build meson

JohnRTitor commented:
@ofborg build ninja

JohnRTitor commented:
> Hard to tell from just this sumary but if this a Musl & coreutils thing see #304856
```
FAIL: tests/shred/shred-exact.sh
# XFAIL: 0
# FAIL:  1
FAIL: tests/shred/shred-exact
FAIL tests/shred/shred-exact.sh (exit status: 1)
# XFAIL: 0
# FAIL:  1
```

@LeSuisse Looks like this exact test is failing.

JohnRTitor commented:
@ofborg test coreutils

vcunat reviewed (changes requested):
```
error: hash mismatch in fixed-output derivation '/nix/store/13zfv027g2bzjdwkcra1f2a72ap59md0-allow-building-with-ninja-1.12.diff.drv':
         specified: sha256-2rlIQHXGox2eeDzqZQz1TYYUqLDhrDqGXHusM8y6+LY=
            got:    sha256-Y+uEXomtYcLQ8tXnopceh/ZGUpYCSc6oJUkbyRpg9HY=
```

I see you used hash for the file before `fetchpatch` edited it (which is the difference from `fetchurl`).

JohnRTitor commented:
> I see you used hash for the file before `fetchpatch` edited it (which is the difference from `fetchurl`).

@vcunat Thanks for the fix!
I am having trouble locally building meson, coreutils check is failing for some reason locally

And ofborg can't build as it can't apply the commit patch. Weird. Any sugggestions will be helpful.

vcunat reviewed (changes requested):
Meson won't build (tested with this PR based on nixpkgs master where meson version is the same as on staging):
```
error: builder for '/nix/store/6qdwjawd7vakslwblggr3yns5xlks584-meson-1.4.0.drv' failed with exit code 1;
       last 10 log lines:
       > patching file 'test cases/rust/18 proc-macro/transitive-proc-macro.rs'
       > applying patch /nix/store/kq5czaf589y55rnwi1i5fbbpbrgk10kd-allow-building-with-ninja-1.12.diff
       > patching file run_project_tests.py
       > Hunk #2 succeeded at 720 (offset -3 lines).
       > patching file run_tests.py
       > patching file unittests/baseplatformtests.py
       > Hunk #1 succeeded at 28 (offset -3 lines).
       > Hunk #2 FAILED at 294.
       > Hunk #3 succeeded at 304 (offset -5 lines).
       > 1 out of 3 hunks FAILED -- saving rejects to file unittests/baseplatformtests.py.rej
       For full logs, run 'nix log /nix/store/6qdwjawd7vakslwblggr3yns5xlks584-meson-1.4.0.drv'.
```

JohnRTitor commented:
The thing is I see no such errors, when I run `nix build .#meson` it just proceeds to download deps and starts building then fails at coreutils checkphase, not at meson patchphase. Tested from this branch `ninja`.

vcunat commented:
coreutils is way too early, so you won't get near trying to build meson.

JohnRTitor commented:
> coreutils is way too early, so you won't get near trying to build meson.

So, what would be the solution to this problem? How do I safely apply this patch? It fails in nix CI, as you can see.

JohnRTitor committed:
meson: add patch to allow building via ninja 1.12
adapted from https://github.com/mesonbuild/meson/commit/1b9eb6f3134d61c5ea0026ac95d807aab812c414

JohnRTitor commented:
@ofborg eval
@ofborg build meson
@ofborg build ninja

JohnRTitor commented:
@vcunat @NickCao  ofborg was able to successfully build meson with the patch.

vcunat commented:
At this point I'd rather wait a while after NixOS 24.05 is forked off.

AndersonTorres commented on pkgs/by-name/me/meson/package.nix:
```diff
@@ -69,6 +69,10 @@ python3.pkgs.buildPythonApplication rec {
     # https://github.com/mesonbuild/meson/issues/12973
     ./0001-Revert-rust-recursively-pull-proc-macro-dependencies.patch
+
+    # Fix compilation of Meson using Ninja 1.12
```
Why vendoring this patch?

JohnRTitor replied:
Because fetching the patch directly from https://github.com/mesonbuild/meson/commit/1b9eb6f3134d61c5ea0026ac95d807aab812c414 won't work as it depends on previous commits.

The issue Is exactly with patching file `unittests/baseplatformtests.py`. As you can see in the above discussion.

To ease the patching process, I manually cherry-picked and squashed the changes into a patch.

JohnRTitor committed:
ninja: 1.12.0 -> 1.12.1
add nix-update-script
move to finalAttrs

JohnRTitor committed:
ninja: migrate to pkgs/by-name

JohnRTitor commented:
@ofborg eval
@ofborg build ninja
@ofborg build meson

JohnRTitor commented:
@ofborg eval
@ofborg build ninja
@ofborg build meson

JohnRTitor commented:
@vcunat since 24.05 have been branched-off already, I guess it's time to move this PR forward?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3981

vcunat reviewed (approved):
Looks good, `meson` builds for me.  News best at: https://github.com/ninja-build/ninja/releases
